asciidoc/client-applications.adoc

```diff
@@ -63,87 +57,67 @@ include::{javascript-examples}/examples.test.js[tags=driver-lifecycle]
-----
-======
+// [.include-with-python]
+// ======
+// [source, python]
+// ----
+// include::{python-examples}/driver_lifecycle_example.py[tags=driver-lifecycle]
+// ----
+// ======
 ====
 
 
 [[driver-connection-uris]]
 == Connection URIs
 
-A connection URI identifies a graph database and how to connect to it.
-The official Neo4j drivers currently support the following URI schemes and driver object types:
+Connection URIs are typically formed according to the following pattern:
 
-.Available URI schemes
-[options="header"]
-|===
-| URI Scheme     | Driver object type
-| bolt           | Direct driver
-| bolt+routing   | Routing driver
-|===
+----
+neo4j://<HOST>:<PORT>[?<ROUTING_CONTEXT>]
+----
 
+This targets a routed Neo4j service that may be fulfilled by either a cluster or a single instance.
+The HOST and PORT values contain a logical hostname and port number targeting the entry point to the Neo4j service (e.g. neo4j://graph.example.com:7687).
+In a clustered environment, this will resolve to one of more of the core members; for standalone installations, this will simply point to that server address.
+The ROUTING_CONTEXT option allows for customization of the routing table and is discussed in more detail in <<driver-routing-context>>.
 
-=== Direct drivers (bolt)
+An alternative URI form, using the bolt URI scheme (e.g. `bolt://graph.example.com:7687`), can be used when a single point-to-point connection is required.
+This variant is useful for the subset client applications (such as admin tooling) that need to be aware of individual servers, as opposed to those which require a highly available database service.
 
-A direct driver is created via a `bolt` URI, for example: `bolt://localhost:7687`.
-This kind of driver is used to maintain connections to a single database server and is typically used when working with a single Neo4j instance or when targeting a specific member of a cluster.
-Note that a routing driver is generally preferable when working with a Causal Cluster.
+----
+bolt://<HOST>:<PORT>
+----
 
+=== Initial address resolution
 
-[[routing_drivers_bolt_routing]]
-=== Routing drivers (bolt+routing)
+The address provided in a `neo4j:// URI` is used for initial and fallback communication only.
+This communication occurs to bootstrap the initial routing table, through which all subsequent communication is carried out.
+Fallback occurs when the driver is unable to contact any of the addresses held in the routing table, and the initial address is once again reused to bootstrap the system.
 
-A routing driver is created via a `bolt+routing` URI, for example: `bolt+routing://graph.example.com:7687`.
-The address in the URI must be that of a Core Server.
-This kind of driver uses the <<term-bolt-routing-protocol, Bolt Routing Protocol>> and works in tandem with the cluster to route transactions to available cluster members.
+Several options are available for providing this initial logical to physical host resolution.
```

Comment:
"logica-to-physical host resolution"?

Reply:
I agree! "logical-to-physical" it shall be.